vercel/vercel
# [node-bridge][node][build-utils] add support for isomorphic signature serverless functions

### 📖 What's in there?

Introduces support for web-compliant signatures (aka isomorphic) `(request: Request) => MaybePromise<Response>` signature in serverless functions.

This PR covers prebuilt apps (Build Output API documentation update to come) and Vercel applications (both locally and in production, tested with build-container).

[RFC](https://www.notion.so/vercel/Isomorphic-Functions-part-1-0e89cf379f79445086112f09659bb978#ba806fed7d1b4a828a2b2a82c334251f)

This covers:
- feat(node-bridge): based on configured `launcherType: WinterCG-experimental` property, wraps the user-provided isomorphic function into a node.js compliant function
- feat(node): sets `launcherType` based on explicit runtime configured in the user function (`export const config = { runtime: 'nodejs-experimental' }`), both when building and developing
- feat(node): throws error when developing when executing an isomorphic serverless function on node <18
- feat(node): throws error while building when the targeted node.js version of an isomorphic serverless function is <18

### 🧪 How to test locally

One can set a local vc app and build it locally.
1. build all packages locally `pnpm i && pnpm build`
1. create a new folder with this content:
   ```sh
   // file api/serverless.js:
   export default async function handler() {
     return Response.json({ works: 'just right from serverless' });
   }

   export const config = { runtime: 'nodejs-experimental' }
   ```
1. run it locally with node 18+: `<path to vercel repo>/packages/cli/dist/index.js dev` and browse to localhost:3000/api/serverless
   > it should serve the json payload
1. run it locally with node <18: `<path to vercel repo>/packages/cli/dist/index.js dev` and browse to localhost:3000/api/serverless
   > the request fails and the appropriate message is displayed in terminal
1. build your app: `<path to vercel repo>/packages/cli/dist/index.js build`
   > inspect `vercel/output/functions/api/serverless.func/.vc-config.json`: its launcher type should be `WinterCG-experimental`
1. add a package.json file to your app with `"engines": { "node": "16.x.x" }` and rebuild 
   > build will fails and the appropriate message on incompatibility
